addons/mail/static/src/core/common/thread_model.js

```diff
@@ -587,6 +590,49 @@ export class Thread extends Record {
         return this.memberCount - this.channelMembers.length;
     }
 
+    isReadonly(fetchAccessRights = true) {
+        if (this.model === "discuss.channel" || !this.id) {
+            return false;
+        }
+        if (this.hasWriteAccess === undefined && fetchAccessRights) {
+            sequential(async () => {
+                if (this.hasWriteAccess !== undefined) {
+                    return;
+                }
+                const threadData = {};
+                const mailBoxThread = this.store.discuss.thread;
+                if (this.store.discuss.isActive && mailBoxThread.model === "mail.box") {
+                    for (const message of mailBoxThread.nonEmptyMessages) {
+                        if (
+                            message.thread &&
+                            message.thread.hasWriteAccess === undefined &&
+                            message.thread.model !== "discuss.channel" &&
+                            message.thread.id
+                        ) {
+                            if (!threadData[message.thread.model]) {
+                                threadData[message.thread.model] = new Set();
+                            }
+                            threadData[message.thread.model].add(message.thread.id);
+                        }
+                    }
+                    for (const model in threadData) {
+                        threadData[model] = Array.from(threadData[model]);
+                    }
+                } else {
+                    threadData[this.model] = [this.id];
+                }
+                if (Object.keys(threadData).length > 0) {
+                    const res = await rpc("/mail/thread/data/access_rights", {
+                        thread_data: threadData,
+                    });
+                    this.store.insert(res);
+                }
+            });
+            return false;
+        }
+        return !this.hasWriteAccess && !(this.hasReadAccess && this.canPostOnReadonly);
```

Comment:
`useSequential` has the feature to batch in single RPC for many requests:

- the `useSequential` should be owned by proper context. It's not on a thread but a group of threads. So need to move to a looser context, e.g. the `store` model;
- Inner-code of `useSequential` should not depend on `this` of thread. This inner function can run once for many inputs. The pattern is to register the callers data before `sequential()`, and the first LOCs of `sequential` inner-function is to process these requests in queue. It early returns if there are not relevant requests to process, otherwise it goes on.

So this is how it should be used:
- each call of `thread.isReadonly/canPost` registers this thread for fetching if we don't know the access rights of it (and it's not discuss.channel which we can deduced easily)
- The 1st thing the inner function should check is to empty queue of registered threads and filter the threads that need access rights fetching (some could have received them in the meantime). If there are no threads with missing access rights info, don't RPC. Otherwise RPCs on these selected threads, then insert the data of their access rights afterwards.

Reply:
All of this logic should just be hidden inside `thread.fetchData` which itself should use `store.fetchData` which already implements batching in a single query.

Reply:
IMO `sequential` is better here rather than `debounced`.

Reply:
> IMO `sequential` is better here rather than `debounced`.

the debounce is 1ms, effectively it just batches synchronous and immediately resolved asynchronous code. Inner-code of `store.fetchData` doesn't use `sequential` but their code is quite similar and it could have used `sequential`, so you don't really loose anything with using `store.fetchData`

Reply:
I tried doing the thread data fetching using `store.fetchData`. But, it doesn't work as expected. As, `1ms` delay is too short it doesn't batch all the calls made to `fetchData` because of `hasWriteAcess`. Which in the end ends up making multiple calls to `mail/data`.

Reply:
In which flow is 1ms delay too short? Data from the same batch should be processed synchronously.

Reply:
These multiple calls are happening when we open a document's chatter. When it is being loaded the `readOnly` method would get called quite a lot of time form re-renders. In my test run it was being called around 15-20 times. As, `readOnly` has some computations to do, it makes a slight delay between subsequent `fetchData` calls.

Because of this it ends up making 4-5 RPC calls. But, this gets resolved when I increase the FETCH_DATA_DELAY.